python/ray/tune/sample.py

```diff
@@ -1,97 +1,408 @@
 import logging
 import random
+from copy import copy
+from numbers import Number
+from typing import Any, Callable, Dict, List, Optional, Sequence, Union
 
 import numpy as np
 
 logger = logging.getLogger(__name__)
 
 
-class sample_from:
+class Domain:
+    sampler = None
+    default_sampler_cls = None
+
+    def cast(self, value):
+        """Cast value to domain type"""
+        return value
+
+    def set_sampler(self, sampler, allow_override=False):
+        if self.sampler and not allow_override:
+            raise ValueError("You can only choose one sampler for parameter "
+                             "domains. Existing sampler for parameter {}: "
+                             "{}. Tried to add {}".format(
+                                 self.__class__.__name__, self.sampler,
+                                 sampler))
+        self.sampler = sampler
+
+    def get_sampler(self):
+        sampler = self.sampler
+        if not sampler:
+            sampler = self.default_sampler_cls()
+        return sampler
+
+    def sample(self, spec=None, size=1):
+        sampler = self.get_sampler()
+        return sampler.sample(self, spec=spec, size=size)
+
+    def is_grid(self):
+        return isinstance(self.sampler, Grid)
+
+    def is_function(self):
+        return False
+
+
+class Sampler:
+    def sample(self,
+               domain: Domain,
+               spec: Optional[Union[List[Dict], Dict]] = None,
+               size: int = 1):
+        raise NotImplementedError
+
+
+class BaseSampler(Sampler):
+    def __str__(self):
+        return "Base"
+
+
+class Uniform(Sampler):
+    def __str__(self):
+        return "Uniform"
+
+
+class LogUniform(Sampler):
+    def __init__(self, base: float = 10):
+        self.base = base
+        assert self.base > 0, "Base has to be strictly greater than 0"
+
+    def __str__(self):
+        return "LogUniform"
+
+
+class Normal(Sampler):
+    def __init__(self, mean: float = 0., sd: float = 0.):
+        self.mean = mean
+        self.sd = sd
+
+        assert self.sd > 0, "SD has to be strictly greater than 0"
+
+    def __str__(self):
+        return "Normal"
+
+
+class Grid(Sampler):
+    """Dummy sampler used for grid search"""
+
+    def sample(self,
+               domain: Domain,
+               spec: Optional[Union[List[Dict], Dict]] = None,
+               size: int = 1):
+        return RuntimeError("Do not call `sample()` on grid.")
+
+
+class Float(Domain):
+    class _Uniform(Uniform):
+        def sample(self,
+                   domain: "Float",
+                   spec: Optional[Union[List[Dict], Dict]] = None,
+                   size: int = 1):
```

Comment:
iirc `size` isn't actually used normally right?

Reply:
Only in tests currently, where it's quite handy to check for distribution properties. However I could just sample several times in the tests.
It feels kind of natural to have that parameter for a sample method, but I agree that we currently do not expose it to users and could thus remove it for now.

Reply:
IMO we don't need `size` but if we decide to keep it we should make it a shape of the return tensor (e.g., tuple of int) to have some consistency with `torch.distributions`, `np.random`, `tensorflow_probability`, etc.